contrib/pg_tde/src/access/pg_tde_tdemap.c

```diff
@@ -708,6 +706,22 @@ pg_tde_perform_rotate_key(TDEPrincipalKey *principal_key, TDEPrincipalKey *new_p
 	old_fd = pg_tde_open_file_read(old_path, &old_curr_pos);
 	new_fd = keyrotation_init_file(&new_signed_key_info, new_path, old_path, &new_curr_pos);
 
+	if (write_xlog)
+	{
+		XLogPrincipalKeyRotate *xlrec = palloc_object(XLogPrincipalKeyRotate);
```

Comment:
This could be allocated on the stack. The data is copied on `XlogInsert()`.

Reply:
Also there is no reason to generate it this early. Wouldn't it make more sense to generated it exactly when we did it? Because otherwise we rotate of the replica even if the old file is corrupted on the master.

Reply:
I personally feel like the middle was a good place. :D

Reply:
Having it at the same place, we can end up with a rotating replica's key even if the primary rotation hasn't happened. Then, it doesn't differ much from Xlogging beforehand, except that it has fewer chances to happen, but the possibility is still there.

Reply:
Good point. I am almost convinced now that it is totally fine the "write behind log" this. And PostgreSQL already does similar things in for example the SMGR code where it creates the file first and then logs it later. Need to mull on this a bit more fo the future but I think we can merge the change as-is.

Reply:
What you should do is add a comment of why this behviour is sane so future people can figure out why it was done.

Reply:
I've added a comment.
Btw, if to look at Heap AM insert or update code, they first modify the buffer, and only after do XLogInsert(). Despite the sequence of those actions inside the function, the only guarantee is that the WAL with those records will be flushed to disk before backed returns "transaction committed". And that will happen well after the function with XLogInsert() returns.